compiler/src/dmd/lexer.d

```diff
@@ -701,12 +704,102 @@ class Lexer
+
+                        while(priorVPtr < &priorValidation[$-1])
+                        {
+                            if (*priorVPtr == '\\')
+                            {
+                                dchar tempDchar = 0;
+                                uint times;
+
+                                // universal character name (C)
+                                if (priorVPtr[1] == 'u')
+                                    times = 4;
+                                else if (priorVPtr[1] == 'U')
+                                    times = 8;
+                                else
+                                    assert(0, "ICE: Universal character name is 2 or 4 bytes only");
+                                priorVPtr += 2;
+
+                                foreach(i; 0 .. times)
+                                {
+                                    char c = priorVPtr[i];
+
+                                    if (c >= '0' && c <= '9')
+                                        c -= '0';
+                                    else if (c >= 'a' && c <= 'f')
+                                        c -= 'a' - 10;
+                                    else if (c >= 'A' && c <= 'F')
+                                        c -= 'A' - 10;
+
+                                    tempDchar <<= 4;
+                                    tempDchar |= c;
+                                }
+
+                                priorVPtr += times - 1; // skip to last character in the hex value
+
+                                if (!utf_isValidDchar(tempDchar))
+                                {
+                                    error(t.loc, "char 0x%x code point is not allowed in an identifier", tempDchar);
+                                }
+
+                                utf_encodeChar(&storage[offset], tempDchar);
+                                offset += utf_codeLengthChar(tempDchar);
+
+                                // Could be an error instead of a warning,
+                                //  but hey it was written specifically so why worry?
+                                if (priorVPtr is priorValidation.ptr)
+                                {
+                                    if (!charLookup.isStart(tempDchar))
+                                        warning(t.loc, "char 0x%x is not allowed start character for an identifier", tempDchar);
+                                }
+                                else
+                                {
+                                    if (!charLookup.isContinue(tempDchar))
+                                        warning(t.loc, "char 0x%x is not allowed continue character for an identifier", tempDchar);
+                                }
+                            }
+                            else
+                            {
+                                storage[offset++] = *priorVPtr;
+                                priorVPtr++;
+                            }
```

Comment:
By the way, what would happen if there's mixed ucn and unicode characters in the identifier symbol?

Reply:
Once this PR is done, if you want to rewrite you can.

Gotta get it working first.